File: services/api/app/api/v1/schemas/rating.py
from pydantic import BaseModel, Field


class CreateRatingRequest(BaseModel):
    movie_id: str
    score: float = Field(ge=0.0, le=10.0, description="Rating on 10-point scale (0.0-10.0)")


class UpdateRatingRequest(BaseModel):
    score: float = Field(ge=0.0, le=10.0)


class RatingResponse(BaseModel):
    id: str
    user_id: str
    movie_id: str
    score: float
    created_at: str
    updated_at: str

    model_config = {"from_attributes": True}


class MovieRatingAggregateResponse(BaseModel):
    avg_score: float
    count: int
    user_rating: float | None = None

File: packages/shared/src/schemas/rating.ts
import { z } from "zod";
import { RATING } from "../constants/rating";

export const RatingSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  movieId: z.string().uuid(),
  score: z.number().min(RATING.MIN).max(RATING.MAX).multipleOf(RATING.STEP),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});

export const CreateRatingSchema = z.object({
  movieId: z.string().uuid(),
  score: z.number().min(RATING.MIN).max(RATING.MAX).multipleOf(RATING.STEP),
});

export const MovieRatingAggregateSchema = z.object({
  avgScore: z.number().min(0).max(10),
  count: z.number().int().min(0),
  userRating: z.number().optional(),
});

export type Rating = z.infer<typeof RatingSchema>;
export type CreateRating = z.infer<typeof CreateRatingSchema>;
export type MovieRatingAggregate = z.infer<typeof MovieRatingAggregateSchema>;
